flare/mgp/mapxb.py

```diff
@@ -0,0 +1,655 @@
+
+global_use_grid_kern = True
+
+class MapXbody:
+    def __init__(self,
+                 grid_num: List,
+                 lower_bound: List or str='auto',
+                 upper_bound: List or str='auto',
+                 svd_rank = 'auto',
+                 species_list: list=[],
+                 map_force: bool=False,
+                 GP: GaussianProcess=None,
+                 mean_only: bool=True,
+                 container_only: bool=True,
+                 lmp_file_name: str='lmp.mgp',
+                 load_grid: str=None,
+                 lower_bound_relax: float=0.1,
+                 n_cpus: int=None,
+                 n_sample: int=100):
+
+        # load all arguments as attributes
+        self.grid_num = np.array(grid_num)
+        self.lower_bound = lower_bound
+        self.upper_bound = upper_bound
+        self.svd_rank = svd_rank
+        self.species_list = species_list
+        self.map_force = map_force
+        self.mean_only = mean_only
+        self.lmp_file_name = lmp_file_name
+        self.load_grid = load_grid
+        self.lower_bound_relax = lower_bound_relax
+        self.n_cpus = n_cpus
+        self.n_sample = n_sample
+        self.spc = []
+        self.spc_set = []
+
+        self.build_bond_struc(species_list)
+
+
+        bounds = [self.lower_bound, self.upper_bound]
+        self.build_map_container(bounds)
+
+        if (not container_only) and (GP is not None) and \
+                (len(GP.training_data) > 0):
+            self.build_map(GP)
+
+    def build_bond_struc(self, species_list):
+        raise NotImplementedError("need to be implemented in child class")
+
+    def get_arrays(self, atom_env):
+        raise NotImplementedError("need to be implemented in child class")
+
+    def build_map_container(self, bounds):
+        '''
+        construct an empty spline container without coefficients.
+        '''
+
+        self.maps = []
+        for spc in self.spc:
+            m = self.singlexbody((self.grid_num, bounds, spc,
+                                  self.map_force, self.svd_rank, self.mean_only,
+                                  self.load_grid, self.lower_bound_relax,
+                                  self.n_cpus, self.n_sample))
+            self.maps.append(m)
+
+
+    def build_map(self, GP):
+        '''
+        generate/load grids and get spline coefficients
+        '''
+
+        self.kernel_info = get_kernel_term(GP, self.kernel_name)
+
+        for m in self.maps:
+            m.build_map(GP)
+
+
+    def predict(self, atom_env, mean_only):
+
+        min_dist = atom_env.bond_array_2[0][0]
+        lower_bound = np.max(self.maps[0].bounds[0][0])
+        assert min_dist >= lower_bound,\
+                f'The minimal distance {min_dist:.3f} is below the mgp lower bound {lower_bound:.3f}'
+
+        if self.mean_only:  # if not build mapping for var
+            mean_only = True
+
+        force_kernel, en_kernel, _, cutoffs, hyps, hyps_mask = self.kernel_info
```

Comment:
should we double check whether hyps_mask is consistent with the cutoffs_mask in atom_env ?